File: 2026/04-April/04-Saturday/Solution.java
//update
class Solution {
    public String decodeCiphertext(String encodedText, int rows) {
        if (rows == 1) return encodedText;

        int n = encodedText.length();
        int cols = n / rows;

        char[][] matrix = new char[rows][cols];
        int idx = 0;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                matrix[i][j] = encodedText.charAt(idx++);
            }
        }

        StringBuilder res = new StringBuilder();

        for (int start = 0; start < cols; start++) {
            int i = 0, j = start;
            while (i < rows && j < cols) {
                res.append(matrix[i][j]);
                i++;
                j++;
            }
        }

        return res.toString().replaceAll("\\s+$", "");
    }
}

File: 2026/04-April/04-Saturday/Solution.kt
//update
class Solution {
    fun decodeCiphertext(encodedText: String, rows: Int): String {
        if (rows == 1) return encodedText

        val n = encodedText.length
        val cols = n / rows

        val matrix = Array(rows) { CharArray(cols) }
        var idx = 0

        for (i in 0 until rows) {
            for (j in 0 until cols) {
                matrix[i][j] = encodedText[idx++]
            }
        }

        val res = StringBuilder()

        for (start in 0 until cols) {
            var i = 0
            var j = start
            while (i < rows && j < cols) {
                res.append(matrix[i][j])
                i++
                j++
            }
        }

        return res.toString().trimEnd()
    }
}
